slick/src/main/scala-3/slick/lifted/ShapedValue.scala

```diff
@@ -69,10 +69,14 @@ object ShapedValue {
         }
     }
 
-    if(elemTpes.length != targetElemTpes.length) {
+    val notMatchTypes = (elemTpes.length > 1) && (
+      !elemTpes.lazyZip(targetElemTpes).forall((x, y) => TypeRepr.of(using x) =:= TypeRepr.of(using y))
+    )
+
+    if ((elemTpes.length != targetElemTpes.length) || notMatchTypes) {
       // todo: change
-      val src = elemTpes.iterator.map(x => Type.show(using summon[Type[U]])).mkString("(", ", ", ")")
-      val target = targetElemTpes.iterator.map(x => Type.show(using summon[Type[U]])).mkString("(", ", ", ")")
+      val src = elemTpes.iterator.map(x => Type.show(using x)).mkString("(", ", ", ")")
+      val target = targetElemTpes.iterator.map(x => Type.show(using x)).mkString("(", ", ", ")")
```

Comment:
I believe `src` and `target` are the wrong way round in your error message

Reply:
Perhaps the message could be more user friendly anyway. Maybe something like this?

```
  |          Can not map to Coffee because the fields don't match.
  |            Mapping from: (scala.Double, java.lang.String)
  |            To:           (java.lang.String, scala.Double)
```